app/lib/job/backend.dart

```diff
@@ -0,0 +1,275 @@
+        tx.queueMutations(inserts: [current]);
+        await tx.commit();
+        return;
+      } else {
+        _logger.info('Creating job: $id');
+        final job = new Job()
+          ..id = id
+          ..service = service
+          ..packageName = package
+          ..packageVersion = version
+          ..isLatestStable = isLatestStable
+          ..packageVersionUpdated = updated
+          ..state = state
+          ..lockedUntil = lockedUntil
+          ..lastStatus = JobStatus.none
+          ..runtimeVersion = versions.runtimeVersion
+          ..errorCount = 0
+          ..updatePriority();
+        tx.queueMutations(inserts: [job]);
+        await tx.commit();
+        return;
+      }
+    });
+  }
+
+  Future<Job> lockAvailable(JobService service, {Duration lockDuration}) async {
+    final query = _db.query(Job)
+      ..filter('service =', service)
+      ..filter('state =', JobState.available)
+      ..order('priority')
+      ..limit(100);
+    final List<Job> list = await query.run().toList();
+
+    bool isApplicable(Job job) {
+      if (job == null) return false;
+      if (job.state != JobState.available) return false;
+      if (job.runtimeVersion != versions.runtimeVersion) return false;
+      return true;
+    }
+
+    list.removeWhere((job) => !isApplicable(job));
+    if (list.isEmpty) return null;
+
+    final selectedId = list[_random.nextInt(list.length)].id;
+    return _db.withTransaction((tx) async {
+      final items = await tx.lookup([_db.emptyKey.append(Job, id: selectedId)]);
+      final Job selected = items.single;
+      if (!isApplicable(selected)) return null;
+      final now = new DateTime.now().toUtc();
+      selected
+        ..state = JobState.processing
+        ..processingKey = _uuid.v4()
+        ..lockedUntil = now.add(lockDuration ?? _defaultLockDuration);
+      tx.queueMutations(inserts: [selected]);
+      await tx.commit();
+      return selected;
+    });
+  }
+
+  Future unlockStaleProcessing(JobService service) async {
+    Future _unlock(Job job) {
+      return _db.withTransaction((tx) async {
+        final list = await tx.lookup([job.key]);
+        final Job current = list.single;
+        if (current.state == JobState.processing &&
+            current.lockedUntil == job.lockedUntil) {
+          current
+            ..state = JobState.idle
+            ..processingKey = null
+            ..errorCount = current.errorCount + 1
+            ..lastStatus = JobStatus.aborted
+            ..lockedUntil = null
+            ..updatePriority();
+          tx.queueMutations(inserts: [current]);
+          await tx.commit();
+        }
+      });
+    }
+
+    final query = _db.query(Job)
+      ..filter('service =', service)
+      ..filter('state =', JobState.processing)
+      ..filter('lockedUntil <', new DateTime.now().toUtc());
+    await for (Job job in query.run()) {
+      try {
+        if (job.runtimeVersion != versions.runtimeVersion) continue;
```

Comment:
Why don't we want to unlock stale jobs from a different runtime version?

Reply:
removing...